rust/sdk/examples/chat/display.rs

```diff
@@ -52,18 +52,46 @@ pub enum Display {
     Cmdline(Reader<TextArea<'static>>),
     Messages(Reader<Vec<Message>>),
     NotConnected(String),
+    Connected,
     UpdateIdentity(Reader<Identity>),
     Scroll(i8),
-    Connected,
+}
+
+impl Display {
+    pub fn scroll_down(multiplier: i8) -> Self {
+        Display::Scroll(-1 * multiplier)
+    }
+    pub fn scroll_up(multiplier: i8) -> Self {
+        Display::Scroll(1 * multiplier)
+    }
 }
 
 #[derive(Default)]
 struct MessageScrollState {
     /// vertical_scroll_from_the_bottom
-    pub last_recorded_par_height: Option<u16>,
+    pub last_text_height: Option<u16>,
     pub vscroll_from_bottom: u16,
 }
 
+impl MessageScrollState {
+    pub fn update_text_height_and_clamp_vscroll(&mut self, text_height: u16, clamp_max_vscroll: u16) {
+        // if scroll_state.vscroll_from_bottom is not 0, it must shift up, because UX
+        if self.vscroll_from_bottom > 0 {
+            // Is there a height shift from previous render?
+            let height_shift_from_previous_render = self
+                .last_text_height
+                .map_or(0_u16, |prev_height| text_height.saturating_sub(prev_height));
```

Comment:
This means that if the Paragraph got smaller, the scroll offset from bottom stays the same — which is correct if we assume that getting smaller means messages being dropped at the top. If it gets smaller because the terminal width is increased, then we get an effective “scrolling upwards” behaviour. This may be fine, I just wanted to say it.

Reply:
Ah, yeah. I did not catch that scenario